src/Microsoft.AspNetCore.Blazor/Components/ParameterCollection.cs

```diff
@@ -174,5 +174,27 @@ internal bool DefinitelyEquals(ParameterCollection oldParameters)
                 }
             }
         }
+
+        internal void CaptureSnapshot(ArrayBuilder<RenderTreeFrame> builder)
+        {
+            builder.Clear();
+
+            var numEntries = 0;
+            foreach (var entry in this)
+            {
+                numEntries++;
+            }
+
+            // We need to prefix the captured frames with an "owner" frame that
+            // describes the length of the buffer so that ParameterCollection
+            // knows how far to iterate through it.
+            var owner = RenderTreeFrame.PlaceholderChildComponentWithSubtreeLength(1 + numEntries);
+            builder.Append(owner);
+
+            if (numEntries > 0)
+            {
+                builder.Append(_frames, _ownerIndex + 1, numEntries);
+            }
+        }
```

Comment:
I know I'm not super familiar with this code, but I'm totally lost looking at this.... what's this doing and why?

Reply:
Yeah, I'm not super thrilled with this, but it's creating a clone of the parameter collection data that matches the shape of what `ParameterCollectionEnumerator` knows how to step through.

I previously tried making a new data structure but that forced either having two versions of `ParameterCollectionEnumerator` that understands each of them, or actually copying the parameter data to a new buffer even in the non-cascading case. Both seemed worse than this relatively contained bit of copying here.